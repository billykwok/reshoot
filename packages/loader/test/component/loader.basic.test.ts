import { describe, beforeAll, afterAll, test, expect } from '@jest/globals';
import path from 'path';
import SegfaultHandler from 'segfault-handler';
import compile from './compiler';
import createMemfs from './createMemfs';
import { mockBase64, mockMetroHash } from './mock';

SegfaultHandler.registerHandler('crash.basic.log');

describe('Basic component test for @reshoot/loader', () => {
  beforeAll(() => {
    jest.doMock('metrohash', mockMetroHash('tEsThAsH'));
    jest.doMock('../../src/util/base64', mockBase64('data/of=someimage'));
  });

  afterAll((done) => {
    jest.resetModules();
    setTimeout(() => ((done as unknown) as () => void)(), 1000).unref();
  });

  test('should emit files and output successfully with default options', async () => {
    const memfs = createMemfs();
    const actual = await compile(
      memfs,
      '../../../../__fixtures__/test-image.jpg',
      { outputPath: '/images/', cache: false }
    );
    const images = await memfs.promises.readdir(
      path.resolve(__dirname, './images')
    );
    const expectedFilenames = [
      'tEsThAsH-480.avif',
      'tEsThAsH-800.avif',
      'tEsThAsH-1120.avif',
      'tEsThAsH-1440.avif',
      'tEsThAsH-480.webp',
      'tEsThAsH-800.webp',
      'tEsThAsH-1120.webp',
      'tEsThAsH-1440.webp',
      'tEsThAsH-480.jpg',
      'tEsThAsH-800.jpg',
      'tEsThAsH-1120.jpg',
      'tEsThAsH-1440.jpg',
    ];
    expect(images).toEqual(expect.arrayContaining(expectedFilenames));
    expect(images).toHaveLength(expectedFilenames.length);
    expect(actual).toMatchSnapshot();
    memfs.reset();
  }, 120000);

  test('should emit error with invalid options', async () => {
    const memfs = createMemfs();
    const actual = compile(memfs, '../../../../__fixtures__/test-image.jpg', {
      cache: false,
      aspectRatio: { decimal: 11 },
    });
    await expect(actual).rejects.toThrowErrorMatchingSnapshot();
    expect(memfs.existsSync(path.resolve(__dirname, './images'))).toBe(false);
    memfs.reset();
  }, 120000);
});
