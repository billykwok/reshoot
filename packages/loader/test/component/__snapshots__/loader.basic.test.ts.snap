// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic component test for @reshoot/loader should emit error with invalid options 1`] = `undefined`;

exports[`Basic component test for @reshoot/loader should emit files and output successfully with default options 1`] = `"export default {\\"src\\":__webpack_public_path__+\\"/images/tEsThAsH-4774.jpg\\",\\"width\\":4774,\\"height\\":3024,\\"srcSet\\":__webpack_public_path__+\\"/images/tEsThAsH-480.jpg 480w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-640.jpg 640w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-840.jpg 840w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-1080.jpg 1080w\\",\\"aspectRatio\\":1.5787,\\"placeholder\\":\\"data:image/jpeg;base64,data/of=someimage\\",\\"color\\":\\"#283838\\",\\"sources\\":[{\\"type\\":\\"image/webp\\",\\"src\\":__webpack_public_path__+\\"/images/tEsThAsH-4774.webp\\",\\"srcSet\\":__webpack_public_path__+\\"/images/tEsThAsH-480.webp 480w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-640.webp 640w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-840.webp 840w,\\"+__webpack_public_path__+\\"/images/tEsThAsH-1080.webp 1080w\\"}]};"`;
